dm/dm/master/scheduler/scheduler.go

```diff
@@ -674,17 +676,79 @@ func (s *Scheduler) TransferSource(source, worker string) error {
 		}
 		return true
 	})
-
 	if len(runningTasks) > 0 {
-		return terror.ErrSchedulerRequireNotRunning.Generate(runningTasks, source)
+		queryWorkerF := func() (*workerrpc.Response, error) {
+			rpcTimeOut := time.Second * 10 // we relay on ctx.Done() to cancel the rpc, so just set a very long timeout
+			req := &workerrpc.Request{Type: workerrpc.CmdQueryStatus, QueryStatus: &pb.QueryStatusRequest{}}
+			failpoint.Inject("operateQueryWorkerF", func(v failpoint.Value) {
+				resp := &workerrpc.Response{Type: workerrpc.CmdQueryStatus, QueryStatus: &pb.QueryStatusResponse{}}
+				switch v.(string) {
+				case "notInSyncUnit":
+					resp.QueryStatus.SubTaskStatus = append(
+						resp.QueryStatus.SubTaskStatus, &pb.SubTaskStatus{Unit: pb.UnitType_Dump})
+					failpoint.Return(resp, nil)
+				case "allTaskIsPaused":
+					resp.QueryStatus.SubTaskStatus = append(
+						resp.QueryStatus.SubTaskStatus, &pb.SubTaskStatus{Stage: pb.Stage_Paused, Unit: pb.UnitType_Sync})
+					failpoint.Return(resp, nil)
+				default:
+					failpoint.Return(nil, errors.New("query error"))
+				}
+			})
+			return oldWorker.SendRequest(ctx, req, rpcTimeOut)
+		}
+
+		// we only allow automatically transfer-source if all subtasks are in the sync phase.
+		resp, err := queryWorkerF()
+		if err != nil {
+			return terror.Annotatef(err, "failed to query worker: %s status err", oldWorker.baseInfo.Name)
+		}
+		for _, status := range resp.QueryStatus.GetSubTaskStatus() {
+			if status.GetUnit() != pb.UnitType_Sync {
+				return terror.ErrSchedulerRequireRunningTaskInSyncUnit.Generate(runningTasks, source)
+			}
+		}
+
+		// pause this running tasks
+		s.mu.Lock()
+		for _, taskName := range runningTasks {
+			if err := s.UpdateExpectSubTaskStage(pb.Stage_Paused, taskName, source); err != nil {
+				s.mu.Unlock()
+				return err
+			}
+		}
+		s.mu.Unlock()
+
+		// wait all tasks are paused before actually starting scheduling
```

Comment:
maybe reuse `waitOperationOk`? @lichunzhu PTAL

Reply:
now `waitOperationOk` is registered to dm-master `Server` so if want to reuse this func , maybe i need move this func to `scheduler`?